syntax = "proto2";

package prototype.protobuf;

option java_package = "prototype.protobuf";
option java_outer_classname = "Control";

// Bidirectional
message Handshake {
    required int64 magic_number = 1;
}

enum ServerMessageType {
    SUCCESS = 0;
    NAME_TAKEN = 1;
    USER_NOT_FOUND = 2;
    SESSION_STARTING = 3;
    USER_ALREADY_IN_SESSION = 4;
}

// Server to client
// Generic message which has many uses
message ServerMessage {
    required ServerMessageType type = 1;
    required bool success = 2;
    optional string message = 3;
}

// Client to server
message NewUser {
    required string username = 1;
}

// Server to client
message UserList {
    repeated string users = 1;
}

// Client to server
message SessionRequest {
    required string name = 1;
}

// Bidirectional
message SessionClose {
	optional string reason = 1;
}

// Bidirectional
message Heartbeat {
    required int64 utc_time = 1;
}
